### DISCUSSION TITLE
blog/2022-03-04-raspberry-pi-4-i2c-without-drivers

### DISCUSSION THREAD

**Discussion by giscus** [crazychenz/vinnie.work#33] [vinnie.work Comments]
# blog/2022-03-04-raspberry-pi-4-i2c-without-drivers

When working with Raspberry Pi, about 99.999% of the information on the internet is focused on Linux support or other higher level interfaces like the RPi.GPIO python package (which depends on Linux itself). When doing, what I'll snootily refer to as, actual embedded development there usually isn't some first class linux driver to do all the dirty work for you. I ran into this very issue when attempting to drive the RPi4 I2C controller from Uboot without an out of the box U-boot driver for the BCM2711.

https://www.vinnie.work/blog/2022-03-04-raspberry-pi-4-i2c-without-drivers

**Comment by davesliu** [comment]
Thanks for your share.
With your way, does it mean that you control all the bits of I2C including the start, stop, restart, ack and NACK bit of I2C master?
As current BMC2711 of raspberry 4 doesn't support SMBus block read feature, is it possible to suppor this feature with your way?

**Reply by davesliu** [reply]
BTW,  from the manual of BCM2711, its BSC address are as follows.... seems inconsitent with your above example.
• BSC0: 0x7e205000
• BSC1: 0x7e804000
• BSC3: 0x7e205600
• BSC4: 0x7e205800
• BSC5: 0x7e205a80
• BSC6: 0x7e205c00

**Reply by crazychenz** [reply]
Hi @davesliu,

Thanks for checking out the article.

The idea behind bit banging I2C is that our code controls the clock of the I2c peripheral and protocol. If I understand your question correctly, the uboot code above is responsible for the start, stop, restart, ack, and NACK bits of our peripheral. The above example was only a minimal example to show the implementation of a dongle key, I didn't implement a lot of extra error checking or other complex duplexing streams. Also, there is an assumption that the connection between the I2C controller (RPi) and the I2C peripheral (dongle) will not have any interference. I felt this created a good learning opportunity for the students I developed this for to see the effects of an unshielded wire receiving noise (you can trigger this with really long cables near pulsing signals).

The 0x7e000000 vs the 0xfe000000 has to do with the way the memory is configured with the BCM2711. I don't recall the exact details, but there is a section in the BCM2711 manual that goes over the different addressing modes (poorly). In practice, IIRC, I happened upon the 0xfe prefix through some online examples of OpenOCD configurations. I believe I sanity checked it to some degree by looking at the manual, gave it a try, and it worked. .... all part of the fun of embedded systems analysis when you don't have the NDAs or network to get to the _real_ documentation.
